src/compiler/types.ts

```diff
@@ -3005,7 +3005,42 @@ namespace ts {
         isRestParameter?: boolean;
     }
 
-    export type SymbolTable = Map<Symbol>;
+    export type InternalSymbolName =
+        | "__call" // Call signatures
+        | "__constructor" // Constructor implementations
+        | "__new" // Constructor signatures
+        | "__index" // Index signatures
+        | "__export" // Module export * declarations
+        | "__global" // Global self-reference
+        | "__missing" // Indicates missing symbol
+        | "__type" // Anonymous type literal symbol
+        | "__object" // Anonymous object literal declaration
+        | "__jsxAttributes" // Anonymous JSX attributes object literal declaration
+        | "__class" // Unnamed class expression
+        | "__function" // Unnamed function expression
+        | "__computed" // Computed property name declaration with dynamic name
+        | "__resolving__" // Indicator symbol used to mark partially resolved type aliases
+        | "export=" // Export assignment symbol
+        | "default"; // Default export symbol (technically not wholly internal, but included here for usability)
+
+    export type UnderscoreEscapedString = (string & { __escapedIdentifier: void }) | (void & { __escapedIdentifier: void }) | InternalSymbolName;
+
+    /** EscapedStringMap based on ES6 Map interface. */
+    export interface UnderscoreEscapedMap<T> {

```

Comment:
We originally didn't make the key of `Map` generic, so it's hardcoded as `string`. Although maybe a better way to do this would be to define `ESMap<K, V>` and `Map<V> = ESMap<string, V>` and `UnderscoreEscapedMap<V> = ESMap<UnderscoreEscapedString, V>`.